idstore.go

```diff
@@ -0,0 +1,86 @@
+package blockstore
+
+import (
+	"context"
+
+	blocks "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+)
+
+// idstore wraps a BlockStore to add support for identity hashes
+type idstore struct {
+	bs Blockstore
+}
+
+func IdStore(bs Blockstore) Blockstore {
+	return &idstore{bs}
+}
+
+func extractContents(k *cid.Cid) (bool, []byte) {
+	dmh, err := mh.Decode(k.Hash())
+	if err != nil || dmh.Code != mh.ID {
+		return false, nil
+	}
+	return true, dmh.Digest
+}
+
+func (b *idstore) DeleteBlock(k *cid.Cid) error {
+	isId, _ := extractContents(k)
+	if isId {
+		return nil
+	}
+	err := b.bs.DeleteBlock(k)
+	if isId && err == ErrNotFound {
+		return nil
+	}
+	return err
+}
+
+func (b *idstore) Has(k *cid.Cid) (bool, error) {
+	isId, _ := extractContents(k)
+	if isId {
+		return true, nil
+	}
+	return b.bs.Has(k)
+}
+
+func (b *idstore) Get(k *cid.Cid) (blocks.Block, error) {
+	isId, bdata := extractContents(k)
+	if isId {
+		return blocks.NewBlockWithCid(bdata, k)
+	}
+	return b.bs.Get(k)
+}
+
+func (b *idstore) Put(bl blocks.Block) error {
+	isId, _ := extractContents(bl.Cid())
+	if isId {
+		return nil
+	}
+	return b.bs.Put(bl)
+}
+
+func (b *idstore) PutMany(bs []blocks.Block) error {
+	toPut := make([]blocks.Block, 0, len(bs))
+	for _, bl := range bs {
+		isId, _ := extractContents(bl.Cid())
+		if isId {
+			continue
+		}
+		toPut = append(toPut, bl)
+	}
+	err := b.bs.PutMany(toPut)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *idstore) HashOnRead(enabled bool) {
+	b.bs.HashOnRead(enabled)
+}
+
+func (b *idstore) AllKeysChan(ctx context.Context) (<-chan *cid.Cid, error) {
+	return b.bs.AllKeysChan(ctx)
```

Comment:
Interesting issue... we'll need to make sure we don't break GC when adding this feature (GC assumes that this function returns all the blocks we have which it won't do in this case).

Reply:
The entire idea behind identity hashes is that they are not stored so there is no need to GC them.

AllKeysChan is only used on the Sweep pass so I don't see this being an issue.

I also don't see how a pinned id-hash can cause a problem with our current (or any reasonable) implementation.

Reply:
Hm. Yeah, you're right. It doesn't look like an issue.